crates/services/p2p/src/cached_view.rs

```diff
@@ -0,0 +1,319 @@
+use crate::ports::P2pDb;
+use dashmap::DashMap;
+use fuel_core_metrics::p2p_metrics::{
+    increment_p2p_req_res_cache_hits,
+    increment_p2p_req_res_cache_misses,
+};
+use fuel_core_storage::Result as StorageResult;
+use fuel_core_types::{
+    blockchain::SealedBlockHeader,
+    services::p2p::Transactions,
+};
+use std::ops::Range;
+type BlockHeight = u32;
+
+pub struct CachedView {
+    sealed_block_headers: DashMap<BlockHeight, SealedBlockHeader>,
+    transactions_on_blocks: DashMap<BlockHeight, Transactions>,
+    metrics: bool,
+}
+
+impl CachedView {
+    pub fn new(metrics: bool) -> Self {
+        Self {
+            sealed_block_headers: DashMap::new(),
+            transactions_on_blocks: DashMap::new(),
+            metrics,
+        }
+    }
+
+    pub fn clear(&self) {
+        self.sealed_block_headers.clear();
+        self.transactions_on_blocks.clear();
+    }
+
+    fn update_metrics<U>(&self, update_fn: U)
+    where
+        U: FnOnce(),
+    {
+        if self.metrics {
+            update_fn()
+        }
+    }
+
+    pub(crate) fn get_sealed_headers<V>(
+        &self,
+        view: &V,
+        block_height_range: Range<u32>,
+    ) -> StorageResult<Option<Vec<SealedBlockHeader>>>
+    where
+        V: P2pDb,
+    {
+        let mut headers = Vec::new();
+        let mut missing_ranges = block_height_range.clone();
+        for block_height in block_height_range.clone() {
+            if let Some(header) = self.sealed_block_headers.get(&block_height) {
+                headers.push(header.clone());
+            } else {
+                // for the first block not in the cache, start a new range
+                missing_ranges.start = block_height;
+                break;
+            }
+        }
```

Comment:
if you are okay with making the iterator mutable, you can simplify this part of the function:

```suggestion
        let mut headers = Vec::new();
        let mut next_header_to_check = block_height_range.start;

        while let Some(header) = block_height_range
            .next()
            .and_then(|block_height| self.sealed_block_headers.get(&block_height))
        {
            headers.push(header.clone());
            next_header_to_check = next_header_to_check.saturating_add(1);
        }

        // If no headers are missing, this range will be empty.
        let missing_ranges = next_header_to_check..block_height_range.end;

```